Mathlib/Algebra/Homology/Embedding/Boundary.lean

```diff
@@ -0,0 +1,181 @@
+an embedding `embeddingUpIntGE p` of `ComplexShape.up ℕ` in `ComplexShape.up ℤ`
+which sends `n : ℕ` to `p + n`. The (canonical) truncation (`≥ p`) of
+`K : CochainComplex C ℤ` shall be defined as the extension to `ℤ`
+(see `Algebra.Homology.Embedding.Extend`) of
+a certain cochain complex indexed by `ℕ`:
+
+`Q ⟶ K.X (p + 1) ⟶ K.X (p + 2) ⟶ K.X (p + 3) ⟶ ...`
+
+where in degree `0`, the object `Q` identifies to the cokernel
+of `K.X (p - 1) ⟶ K.X p` (this is `K.opcycles p`). In this case,
+we see that the degree `0 : ℕ` needs a particular attention when
+constructing the truncation.
+
+In this file, more generally, for `e : Embedding c c'`, we define
+a predicate `ι → Prop` named `e.BoundaryGE` which shall be relevant
+when constructing the truncation `K.truncGE e` when `e.IsTruncGE`.
+In the case of `embeddingUpIntGE p`, we show that `0 : ℕ` is the
+only element in this lower boundary. Similarly, we define
+`Embedding.BoundaryLE`.
+
+-/
+
+namespace ComplexShape
+
+namespace Embedding
+
+variable {ι ι' : Type*} (c : ComplexShape ι) (c' : ComplexShape ι') (e : Embedding c c')
+
+/-- The lower boundary of an embedding `e : Embedding c c'`, as a predicate on `ι`.
+It is satisfied by `j : ι` when there exists `i' : ι'` not in the image of `e.f`
+such that `c'.Rel i' (e.f j)`. -/
+def BoundaryGE (j : ι) : Prop :=
+  c'.Rel (c'.prev (e.f j)) (e.f j) ∧ ∀ i, ¬c'.Rel (e.f i) (e.f j)
+
+lemma boundaryGE {i' : ι'} {j : ι} (hj : c'.Rel i' (e.f j)) (hi' : ∀ i, e.f i ≠ i') :
+    e.BoundaryGE j := by
+  constructor
+  · simpa only [c'.prev_eq' hj] using hj
+  · intro i hi
+    apply hi' i
+    rw [← c'.prev_eq' hj, c'.prev_eq' hi]
+
+lemma not_boundaryGE_next [e.IsRelIff] {j k : ι} (hk : c.Rel j k) :
+    ¬ e.BoundaryGE k := by
+  dsimp [BoundaryGE]
+  simp only [not_and, not_forall, not_not]
+  intro
+  exact ⟨j, by simpa only [e.rel_iff] using hk⟩
+
+lemma not_boundaryGE_next' [e.IsRelIff] {j k : ι} (hj : ¬ e.BoundaryGE j) (hk : c.next j = k) :
+    ¬ e.BoundaryGE k := by
+  by_cases hjk : c.Rel j k
+  · exact e.not_boundaryGE_next hjk
+  · subst hk
+    simpa only [c.next_eq_self j hjk] using hj
+
+variable {e} in
+lemma BoundaryGE.not_mem {j : ι} (hj : e.BoundaryGE j) {i' : ι'} (hi' : c'.Rel i' (e.f j))
+    (a : ι) : e.f a ≠ i' := fun ha =>
+  hj.2 a (by simpa only [ha] using hi')
+
+lemma prev_f_of_not_boundaryGE [e.IsRelIff] {i j : ι} (hij : c.prev j = i)
+    (hj : ¬ e.BoundaryGE j) :
+    c'.prev (e.f j) = e.f i := by
+  by_cases hij' : c.Rel i j
+  · exact c'.prev_eq' (by simpa only [e.rel_iff] using hij')
+  · obtain rfl : j = i := by
+      simpa only [c.prev_eq_self j (by simpa only [hij] using hij')] using hij
+    apply c'.prev_eq_self
+    intro hj'
+    simp only [BoundaryGE, not_and, not_forall, not_not] at hj
+    obtain ⟨i, hi⟩ := hj hj'
+    rw [e.rel_iff] at hi
+    rw [c.prev_eq' hi] at hij
+    exact hij' (by simpa only [hij] using hi)
+
+variable {e} in
+lemma BoundaryGE.false {j : ι} (hj : e.BoundaryGE j) [e.IsTruncLE] : False := by
```

Comment:
should `isTruncLE` be in the name?

Reply:
Thanks! It is indeed better to include it in the name.